File: cpp/src/VideoFile.cpp
#include "VideoFile.h"
#include <algorithm>
#include <iomanip>
#include <sstream>

VideoFile::VideoFile(const std::string& name, const std::string& path, double sizeMB, const std::string& extension)
    : name(name), path(path), sizeMB(sizeMB), extension(extension) {}

VideoFile VideoFile::fromPath(const fs::path& filePath) {
    std::error_code ec;
    auto fileSize = fs::file_size(filePath, ec);
    
    if (ec) {
        throw std::runtime_error("Could not get file size for: " + filePath.string());
    }
    
    double sizeMB = static_cast<double>(fileSize) / (1024.0 * 1024.0);
    std::string name = filePath.filename().string();
    std::string extension = filePath.extension().string();
    
    // Remove the dot from extension and convert to lowercase
    if (!extension.empty() && extension[0] == '.') {
        extension = extension.substr(1);
    }
    std::transform(extension.begin(), extension.end(), extension.begin(), ::tolower);
    
    return VideoFile(name, filePath.string(), sizeMB, extension);
}

bool VideoFile::isSupportedVideo(const std::string& extension) {
    auto supportedExtensions = getSupportedExtensions();
    std::string lowerExt = extension;
    std::transform(lowerExt.begin(), lowerExt.end(), lowerExt.begin(), ::tolower);
    
    return std::find(supportedExtensions.begin(), supportedExtensions.end(), lowerExt) != supportedExtensions.end();
}

std::vector<std::string> VideoFile::getSupportedExtensions() {
    return {"mp4", "avi", "mov", "mkv", "flv", "wmv", "webm", "m4v", "3gp"};
}

std::string VideoFile::toString() const {
    std::ostringstream oss;
    oss << name << " (" << std::fixed << std::setprecision(1) << sizeMB << " MB)";
    return oss.str();
}

File: rust/src/video_file.rs
use std::fmt;
use std::path::Path;

/// Represents a video file with metadata
#[derive(Debug, Clone)]
pub struct VideoFile {
    pub name: String,
    pub path: String,
    pub size_mb: f64,
    pub extension: String,
}

impl VideoFile {
    /// Create a new VideoFile from a file path
    pub fn new(path: &Path) -> Result<Self, std::io::Error> {
        let metadata = std::fs::metadata(path)?;
        let size_mb = metadata.len() as f64 / (1024.0 * 1024.0);
        
        let name = path
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or("unknown")
            .to_string();
        
        let extension = path
            .extension()
            .and_then(|ext| ext.to_str())
            .unwrap_or("")
            .to_lowercase();
        
        Ok(VideoFile {
            name,
            path: path.to_string_lossy().to_string(),
            size_mb,
            extension,
        })
    }
    
    /// Check if the file has a supported video extension
    pub fn is_supported_video(&self) -> bool {
        const SUPPORTED_EXTENSIONS: &[&str] = &[
            "mp4", "avi", "mov", "mkv", "flv", "wmv", "webm", "m4v", "3gp"
        ];
        
        SUPPORTED_EXTENSIONS.contains(&self.extension.as_str())
    }
}

impl fmt::Display for VideoFile {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} ({:.1} MB)", self.name, self.size_mb)
    }
}
